---
title: "Class 14 RNA-seq Mini-Project"
author: "Joshua Mac"
format: pdf
---
---
title: "RNA-Seq analysis mini-project"
author: "Joshua Mac"
format: pdf
date: Feb 24, 2025
---

**Section 1. Differential Expression Analysis**

```{r}
library(DESeq2)
```

```{r}
metaFile <- "GSE37704_metadata.csv"
countFile <- "GSE37704_featurecounts.csv"

# Import metadata and take a peak
colData = read.csv(metaFile, row.names=1)
head(colData)
```

```{r}
# Import countdata
countData = read.csv(countFile, row.names=1)
head(countData)
```

**Q1. Complete the code below to remove the troublesome first column from countData**

> A1. See code below.

```{r}
# Note we need to remove the odd first $length col
countData <- as.matrix(countData[,-1])
head(countData)
```

**Q2. Complete the code below to filter countData to exclude genes (i.e. rows) where we have 0 read count across all samples (i.e. columns). Tip: What will rowSums() of countData return and how could you use it in this context?**

> A2. See code below.

```{r}
# Filter count data where you have 0 read count across all samples.
countData = countData[rowSums(countData)>0, ]
head(countData)
```
*Running DESeq2*

```{r}
dds = DESeqDataSetFromMatrix(countData=countData,
                             colData=colData,
                             design=~condition)
dds = DESeq(dds)
```

```{r}
dds
```

```{r}
res = results(dds, contrast=c("condition", "hoxa1_kd", "control_sirna"))
```

**Q3. Call the summary() function on your results to get a sense of how many genes are up or down-regulated at the default 0.1 p-value cutoff.**

> A3. See code below.

```{r}
summary(res)
```
*Volcano Plot*

```{r}
plot( res$log2FoldChange, -log(res$padj) )
```

**Q4. Improve this plot by completing the below code, which adds color and axis labels**

> A4. Bet

```{r}
# Make a color vector for all genes
mycols <- rep("gray", nrow(res) )

# Color red the genes with absolute fold change above 2
mycols[ abs(res$log2FoldChange) > 2 ] <- "red"

# Color blue those with adjusted p-value less than 0.01
#  and absolute fold change more than 2
inds <- (res$padj<0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

plot( res$log2FoldChange, -log(res$padj), col=mycols, xlab="Log2(FoldChange)", ylab="-Log(P-value)" )
```

*Gene annotations*

**Q5. Use the mapIDs() function multiple times to add SYMBOL, ENTREZID and GENENAME annotation to our results by completing the code below.**

> A5. Bet, see code below.

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

columns(org.Hs.eg.db)

res$symbol = mapIds(org.Hs.eg.db,
                    keys=row.names(res), 
                    keytype="ENSEMBL",
                    column="SYMBOL",
                    multiVals="first")

res$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")

res$name =   mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="GENENAME",
                    multiVals="first")

head(res, 10)
```

**Q6. Finally for this section let's reorder these results by adjusted p-value and save them to a CSV file in your current project directory.**

> A6. Bet, see below.

```{r}
res = res[order(res$pvalue),]
write.csv(res, file = "deseq_results.csv")
```

**Section 2. Pathway Analysis**

```{r}
# Run in your R console (i.e. not your Rmarkdown doc!)
# BiocManager::install( c("pathview", "gage", "gageData") )

# For old vesrsions of R only (R < 3.5.0)!
#source("http://bioconductor.org/biocLite.R")
#biocLite( c("pathview", "gage", "gageData") )
```

```{r}
library(pathview)
```

```{r}
library(gage)
library(gageData)

data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```

```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

Running:
```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

```{r}
attributes(keggres)
```

```{r}
# Look at the first few down (less) pathways
head(keggres$less)
```

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
# Notice new file in directory
```

```{r}
# A different PDF based output of the same data
pathview(gene.data=foldchanges, pathway.id="hsa04110", kegg.native=FALSE)
# Another new file, but pdf instead of png
```

```{r}
## Focus on top 5 upregulated pathways here for demo purposes only
keggrespathways <- rownames(keggres$greater)[1:5]

# Extract the 8 character long IDs part of each string
keggresids = substr(keggrespathways, start=1, stop=8)
keggresids
```

```{r}
pathview(gene.data=foldchanges, pathway.id=keggresids, species="hsa")
```

**Q7. Can you do the same procedure as above to plot the pathview figures for the top 5 down-reguled pathways?**

> A7. Yes, see below.

```{r}
## Focus on top 5 downregulated pathways here for demo purposes only
keggrespathways_down <- rownames(keggres$less)[1:5]

# Extract the 8 character long IDs part of each string
keggresids_down = substr(keggrespathways_down, start=1, stop=8)
keggresids_down
```

> then (still part of A7)

```{r}
pathview(gene.data=foldchanges, pathway.id=keggresids_down, species="hsa")
```

**Section 3. Gene Ontology (GO)**

```{r}
data(go.sets.hs)
data(go.subs.hs)

# Focus on Biological Process subset of GO
gobpsets = go.sets.hs[go.subs.hs$BP]

gobpres = gage(foldchanges, gsets=gobpsets, same.dir=TRUE)

lapply(gobpres, head)
```

**Section 4. Reactome Analysis**
```{r}
sig_genes <- res[res$padj <= 0.05 & !is.na(res$padj), "symbol"]
print(paste("Total number of significant genes:", length(sig_genes)))
```

```{r}
write.table(sig_genes, file="significant_genes.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)
```

**Q8: What pathway has the most significant “Entities p-value”? Do the most significant pathways listed match your previous KEGG results? What factors could cause differences between the two methods?**

> A8. The "Cell Cycle" pathway has the most significant "Entities p-value" at 1.1E-4. The most significant pathways do not exactly match the KEGG results. While they have cell cycle and DNA replication pathways in common, and maybe RNA transport due to the Reactome's listing of ribosomal machinery related pathways, the KEGG results also emphasize a lot of metabolism and recombination. This ambiguity alone may be a factor causing the differences between the two methods, as the way pathways are defined alters the ordering of significant genes into pathways. Statistical methods and p-value computing seem like possible differences as well.

**Plots for top 5 up- and top 5 down-regulated**

Up-Regulated
```{r}
knitr::include_graphics("hsa04640.png")
knitr::include_graphics("hsa04630.png")
knitr::include_graphics("hsa00140.png")
knitr::include_graphics("hsa04142.png")
knitr::include_graphics("hsa04330.png")
```

Down-Regulated
```{r}
knitr::include_graphics("hsa04110.png")
knitr::include_graphics("hsa03030.png")
knitr::include_graphics("hsa03013.png")
knitr::include_graphics("hsa03440.png")
knitr::include_graphics("hsa04114.png")

```

