---
title: "Class 18 - Pertussis Mini Project"
author: "Joshua Mac"
format: pdf
date: Mar 9, 2025
---

**Pertussis and the CMI-PB project**

*SECTION 1: INVESTIGATING PERTUSSIS CASES BY YEAR*

**Q1. With the help of the R “addin” package datapasta assign the CDC pertussis case number data to a data frame called cdc and use ggplot to make a plot of cases numbers over time.**

> A1. See below.

```{r}
# Install addin "datapasta" package
# install.packages("datapasta") in console

# Assign CDC pertussis case number data to df using datapasta from addins menu
cdc <- data.frame(
                                 Year = c(1922L,1923L,1924L,1925L,
                                          1926L,1927L,1928L,1929L,1930L,1931L,
                                          1932L,1933L,1934L,1935L,1936L,
                                          1937L,1938L,1939L,1940L,1941L,1942L,
                                          1943L,1944L,1945L,1946L,1947L,
                                          1948L,1949L,1950L,1951L,1952L,
                                          1953L,1954L,1955L,1956L,1957L,1958L,
                                          1959L,1960L,1961L,1962L,1963L,
                                          1964L,1965L,1966L,1967L,1968L,1969L,
                                          1970L,1971L,1972L,1973L,1974L,
                                          1975L,1976L,1977L,1978L,1979L,1980L,
                                          1981L,1982L,1983L,1984L,1985L,
                                          1986L,1987L,1988L,1989L,1990L,
                                          1991L,1992L,1993L,1994L,1995L,1996L,
                                          1997L,1998L,1999L,2000L,2001L,
                                          2002L,2003L,2004L,2005L,2006L,2007L,
                                          2008L,2009L,2010L,2011L,2012L,
                                          2013L,2014L,2015L,2016L,2017L,2018L,
                                          2019L,2020L,2021L,2022L),
         No..Reported.Pertussis.Cases = c(107473,164191,165418,152003,
                                          202210,181411,161799,197371,
                                          166914,172559,215343,179135,265269,
                                          180518,147237,214652,227319,103188,
                                          183866,222202,191383,191890,109873,
                                          133792,109860,156517,74715,69479,
                                          120718,68687,45030,37129,60886,
                                          62786,31732,28295,32148,40005,
                                          14809,11468,17749,17135,13005,6799,
                                          7717,9718,4810,3285,4249,3036,
                                          3287,1759,2402,1738,1010,2177,2063,
                                          1623,1730,1248,1895,2463,2276,
                                          3589,4195,2823,3450,4157,4570,
                                          2719,4083,6586,4617,5137,7796,6564,
                                          7405,7298,7867,7580,9771,11647,
                                          25827,25616,15632,10454,13278,
                                          16858,27550,18719,48277,28639,32971,
                                          20762,17972,18975,15609,18617,
                                          6124,2116,3044)
       )
```

Next, plot:
```{r}
library(ggplot2)
cdc_plot<-ggplot(cdc, aes(Year, No..Reported.Pertussis.Cases))+
  geom_point()+
  geom_line()+
  labs(title = "Pertussis Cases by Year (1922-2022)", x = "Year", y = "Number of Cases")
cdc_plot
```

*SECTION 2. A TALE OF TWO VACCINES (wP & aP)*

**Q2. Using the ggplot geom_vline() function add lines to your previous plot for the 1946 introduction of the wP vaccine and the 1996 switch to aP vaccine (see example in the hint below). What do you notice?**

> A2. See below. I notice that with the introduction of the wP vaccine (blue line) in 1946, the number of cases of pertussis drops off significantly due to enhance immunization and preventative protection against the affliction. However, after the 1996 switch to the aP vaccine (red line), we actually see more reported cases of pertussis in the population.

```{r}
cdc_plot + 
  geom_vline(xintercept=1946, color = "blue", linetype="dashed") +
  geom_vline(xintercept=1996, color = "red", linetype="dashed") 
  
```

**Q3. Describe what happened after the introduction of the aP vaccine? Do you have a possible explanation for the observed trend?**

> A3. After the introduction of the aP vaccine, we see cases of pertussis reported rising. Most notably, in 2012, 48,277 cases are reported in the US, which has not been at such levels since 1955 (62786 cases). Clearly a public health concern, we can attribute the trend to possible mutation/evolution of the pertussis bacterium *Bordetella pertussis* and the rising resistance to vaccinations due to antivax communities and spread of misinformation...

*SECTION 3. EXPLORING CMI-PB DATA*

CMI-PB at a glance:
Project that aims to provide to the public long-term humoral and cellular immune response data for those who received either vaccine in infancy with Tdap boosters. It aims to investigate the mechanisms underlying waning protection against pertussis by evaluating the pertussis-specific immune responses overtime as mentioned above.

The CMI-PB API returns JSON data (key-value pairs).

So, we'll need the jsonlite package.

```{r}
# install.packages("jsonlite") in console
library(jsonlite)
```

```{r}
subject <- read_json("https://www.cmi-pb.org/api/subject", simplifyVector = TRUE) 
```

```{r}
head(subject, 3)
```

**Q4. How many aP and wP infancy vaccinated subjects are in the dataset?**

> A4. There are 87 aP infancy vaccinated subjects and 85 wP infancy vaccinated subjects in the dataset. See below.

```{r}
table(subject$infancy_vac)
```

**Q5. How many Male and Female subjects/patients are in the dataset?**

> A5. There are 60 male and 112 female subjects in the dataset. See below.

```{r}
table(subject$biological_sex)
```

**Q6. What is the breakdown of race and biological sex (e.g. number of Asian females, White males etc…)?**

> A6. There is one American Indian/Alaska Native male and no females of that race in the dataset. There are 12 Asian males and 32 Asian females in the dataset. There are three Black or African American males and two Black of African American females in the dataset. There are four males that are of More Than One Race and 15 females that are of More Than One Race in the dataset. There is one male and one female that are both Native Hawaiian or Other Pacific Islander in the dataset. There are seven males and 14 females with races unknown/not reported in the dataset. Finally, there are 32 males and 48 females that consider themselves White in the dataset.

```{r}
table(subject$race, subject$biological_sex)
```

Side-Note for Q7 (the "lubridate" package)
```{r}
# install.packages("lubridate") in console
library(lubridate)
```
```{r}
today() # tells today's date in "YEAR-MM-DD"
```
```{r}
diff<- today() - ymd("2000-01-01") # matching the `today()` format, do this to find how many days since new years 2000
diff
```
```{r}
time_length(diff, "years") # the time difference in years
```

**Q7. Using this approach determine (i) the average age of wP individuals, (ii) the average age of aP individuals; and (iii) are they significantly different?**

> A7. (i) The average age of wP individuals is 36 years, (ii) aP individuals is 27, and (iii) they are significantly different as the p-value of the t-test conducted below is low (<2.2e-16) and less than alpha 0.05, suggesting statistical significance.

```{r}
subject$age <- today() - ymd(subject$year_of_birth)
```

```{r}
library(dplyr)
ap<-subject%>% filter(infancy_vac=="aP")
round(summary(time_length(ap$age, "years")))
```

```{r}
wp<-subject%>%filter(infancy_vac=="wP")
round(summary(time_length(wp$age, "years")))
```

```{r}
t.test(ap$age, wp$age, var.equal = FALSE)
```

**Q8. Determine the age of all individuals at time of boost?**

> A8. See the ages, in years, at time of boost below.

```{r}
days_before_boost<- ymd(subject$date_of_boost) - ymd(subject$year_of_birth)
age_at_boost <- time_length(days_before_boost, "year")
head(age_at_boost, 172)
```

**Q9. With the help of a faceted boxplot or histogram (see below), do you think these two groups are significantly different?**

> A9. We can see visually comparing histograms of the wP and aP datasets that they are different. With the t-test conducted in Q7, we are sure that statisitically they are significantly different.

```{r}
ggplot(subject) +
  aes(time_length(age, "year"),
      fill=as.factor(infancy_vac)) +
         geom_histogram(show.legend=FALSE)+
         facet_wrap(vars(infancy_vac), nrow=2) +
         xlab("Age in years")
```

*Joining multiple tables*

```{r}
# Complete the API URLs...
specimen <- read_json("https://www.cmi-pb.org/api/v5_1/specimen", simplifyVector = TRUE) 
titer <- read_json("https://www.cmi-pb.org/api/v5_1/plasma_ab_titer", simplifyVector = TRUE) 
```

**Q9. Complete the code to join specimen and subject tables to make a new merged data frame containing all specimen records along with their associated subject details:**

> A9. See code below.

```{r}
meta <- left_join(specimen, subject)
dim(meta)
head(meta)
```

**Q10. Now using the same procedure join meta with titer data so we can further analyze this data in terms of time of visit aP/wP, male/female etc.**

> A10. See code below.

```{r}
abdata <- inner_join(titer, meta)
dim(abdata)
```

**Q11. How many specimens (i.e. entries in abdata) do we have for each isotype?**

> A11. For IgE there are 6698 specimens, IgG 7265 specimens, IgG1 11993 specimens, IgG2 12000 specimens, IgG3 12000 specimens, and IgG4 12000 specimens.

```{r}
table(abdata$isotype)
```

**Q12. What are the different $dataset values in abdata and what do you notice about the number of rows for the most “recent” dataset?**

> A12. The different $dataset values in abdata are the number of entries/reports in a given year. The number of rows for the most recent 2023 dataset doubled from the year prior, indicating a 2x jump in cases since the prior year.

```{r}
table(abdata$dataset)
```

*SECTION 4. EXAMINE IgG Ab TITER LEVELS*

```{r}
igg <- abdata %>% filter(isotype == "IgG")
head(igg)
```

**Q13. Complete the following code to make a summary boxplot of Ab titer levels (MFI) for all antigens:**

> A13. See code below.

```{r}
ggplot(igg) +
  aes(MFI_normalised, antigen) +
  geom_boxplot() + 
    xlim(0,75) +
  facet_wrap(vars(visit), nrow=2)
```

**Q14. What antigens show differences in the level of IgG antibody titers recognizing them over time? Why these and not others?**

> A14. The PT and FIM2/3 antigens show increasing levels of IgG antibody titers recognizing them over time. These antigents may have stronger immune responses or longer lasting responses explaining the change over time.

```{r}
ggplot(igg) +
  aes(MFI_normalised, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  facet_wrap(vars(visit), nrow=2) +
  xlim(0,75) +
  theme_bw()
```

```{r}
igg %>% filter(visit != 8) %>%
ggplot() +
  aes(MFI_normalised, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  xlim(0,75) +
  facet_wrap(vars(infancy_vac, visit), nrow=2)
```

**Q15. Filter to pull out only two specific antigens for analysis and create a boxplot for each. You can chose any you like. Below I picked a “control” antigen (“OVA”, that is not in our vaccines) and a clear antigen of interest (“PT”, Pertussis Toxin, one of the key virulence factors produced by the bacterium B. pertussis).**

> A15. See below.

```{r}
filter(igg, antigen=="OVA") %>%
  ggplot() +
  aes(MFI_normalised, col=infancy_vac) +
  geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  theme_bw()
```

```{r}
filter(igg, antigen=="PT") %>%
  ggplot() +
  aes(MFI_normalised, col=infancy_vac) +
  geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  theme_bw()
```

**Q16. What do you notice about these two antigens time courses and the PT data in particular?**

> A16. Keeping the difference in axes in mind, we can see that PT levels rise and exceed OVA values before peaking at visit 5 and declining again. wP (blue) and aP (red) individuals show similar trends in PT levels, but after visit 9 in wP individuals, OVA levels decline while they do not in aP individuals.

**Q17. Do you see any clear difference in aP vs. wP responses?**

> A17. The wP group experiences higher levels of PT antigen compared to aP and as mentioned in Q16, OVA antigen levels start to decline after visit 9 in the wP group while it remains a constant level after visit 9.

```{r}
abdata.21 <- abdata %>% filter(dataset == "2021_dataset")

abdata.21 %>% 
  filter(isotype == "IgG",  antigen == "PT") %>%
  ggplot() +
    aes(x=planned_day_relative_to_boost,
        y=MFI_normalised,
        col=infancy_vac,
        group=subject_id) +
    geom_point() +
    geom_line() +
    geom_vline(xintercept=0, linetype="dashed") +
    geom_vline(xintercept=14, linetype="dashed") +
  labs(title="2021 dataset IgG PT",
       subtitle = "Dashed lines indicate day 0 (pre-boost) and 14 (apparent peak levels)")
```

**Q18. Does this trend look similar for the 2020 dataset?**

> A18. Let's look! Not quite, it appears that PT levels in aP individuals showed higher IgG antibody titers recognizing them over time than the 2021 data set.

```{r}
abdata.21 <- abdata %>% filter(dataset == "2020_dataset")

abdata.21 %>% 
  filter(isotype == "IgG",  antigen == "PT") %>%
  ggplot() +
    aes(x=planned_day_relative_to_boost,
        y=MFI_normalised,
        col=infancy_vac,
        group=subject_id) +
    geom_point() +
    geom_line() +
    geom_vline(xintercept=0, linetype="dashed") +
    geom_vline(xintercept=14, linetype="dashed") +
  labs(title="2020 dataset IgG PT",
       subtitle = "Dashed lines indicate day 0 (pre-boost) and 14 (apparent peak levels)")
```

*5. OBTAINING CMI-PB RNASEQ DATA*

```{r}
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"

rna <- read_json(url, simplifyVector = TRUE) 
```

```{r}
#meta <- inner_join(specimen, subject) in console
ssrna <- inner_join(rna, meta)
```

**Q19. Make a plot of the time course of gene expression for IGHG1 gene (i.e. a plot of visit vs. tpm).**

> A19. See code.

```{r}
ggplot(ssrna) +
  aes(visit, tpm, group=subject_id) +
  geom_point() +
  geom_line(alpha=0.2)
```

**Q20. What do you notice about the expression of this gene (i.e. when is it at it’s maximum level)?**

> A20. Expression of this gene is at it's maximum level at visit 4 and then steadly drops in expression afterwards. This is when the body has generated the highest amount of antibodies.

**Q21. Does this pattern in time match the trend of antibody titer data? If not, why not?**

> A21. This pattern in time does match the trend of antibody titer data, as the expression of the gene creates the antibodies that we saw eventually decreasing MFI when we analyzed antigen levels per visit.

```{r}
ggplot(ssrna) +
  aes(tpm, col=infancy_vac) +
  geom_boxplot() +
  facet_wrap(vars(visit))
```
```{r}
ssrna %>%  
  filter(visit==4) %>% 
  ggplot() +
    aes(tpm, col=infancy_vac) + geom_density() + 
    geom_rug() 
```



